tf5muxserver/mux_server_ListResource.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package tf5muxserver
+
+import (
+	"context"
+	"slices"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5"
+
+	"github.com/hashicorp/terraform-plugin-mux/internal/logging"
+)
+
+func (s *muxServer) ListResource(ctx context.Context, req *tfprotov5.ListResourceRequest) (*tfprotov5.ListResourceServerStream, error) {
+	rpc := "ListResource"
+	ctx = logging.InitContext(ctx)
+	ctx = logging.RpcContext(ctx, rpc)
+
+	server, diags, err := s.getListResourceServer(ctx, req.TypeName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// slices.all package
+
+	if diagnosticsHasError(diags) {
+		return &tfprotov5.ListResourceServerStream{
+			Results: slices.Values([]tfprotov5.ListResourceResult{
+				{
+					Diagnostics: diags,
+				},
+			}),
+		}, nil
+	}
+
+	// TODO: Remove and call server.ListResource below directly once interface becomes required.
+	//nolint:staticcheck // Intentionally verifying interface implementation
+	listResourceServer, ok := server.(tfprotov5.ProviderServerWithListResource)
+	if !ok {
+		resp := &tfprotov5.ListResourceServerStream{
+			Results: slices.Values([]tfprotov5.ListResourceResult{
+				{
+					Diagnostics: []*tfprotov5.Diagnostic{
+						{
+							Severity: tfprotov5.DiagnosticSeverityError,
+							Summary:  "ListResource Not Implemented",
+							Detail: "A ListResource call was received by the provider, however the provider does not implement ListResource. " +
+								"Either upgrade the provider to a version that implements ListResource or this is a bug in Terraform that should be reported to the Terraform maintainers.",
```

Comment:
💭 Not specific to this pull request: I wonder how this could be `a bug in Terraform`.